cpp/core/utils/metrics.h

```diff
@@ -17,104 +17,72 @@
 
 #pragma once
 
+#include <memory>
+
 namespace gluten {
 
 struct Metrics {
-  int numMetrics = 0;
-
-  long* inputRows;
-  long* inputVectors;
-  long* inputBytes;
-  long* rawInputRows;
-  long* rawInputBytes;
-  long* outputRows;
-  long* outputVectors;
-  long* outputBytes;
-
-  // CpuWallTiming.
-  long* cpuCount;
-  long* wallNanos;
-  long veloxToArrow;
-
-  long* peakMemoryBytes;
-  long* numMemoryAllocations;
-
-  // Spill
-  long* spilledBytes;
-  long* spilledRows;
-  long* spilledPartitions;
-  long* spilledFiles;
-
-  // Runtime metrics.
-  long* numDynamicFiltersProduced;
-  long* numDynamicFiltersAccepted;
-  long* numReplacedWithDynamicFilterRows;
-  long* flushRowCount;
-  long* scanTime;
-  long* skippedSplits;
-  long* processedSplits;
-  long* skippedStrides;
-  long* processedStrides;
-
-  Metrics(int size) : numMetrics(size) {
-    inputRows = new long[numMetrics]();
-    inputVectors = new long[numMetrics]();
-    inputBytes = new long[numMetrics]();
-    rawInputRows = new long[numMetrics]();
-    rawInputBytes = new long[numMetrics]();
-    outputRows = new long[numMetrics]();
-    outputVectors = new long[numMetrics]();
-    outputBytes = new long[numMetrics]();
-    cpuCount = new long[numMetrics]();
-    wallNanos = new long[numMetrics]();
-    peakMemoryBytes = new long[numMetrics]();
-    numMemoryAllocations = new long[numMetrics]();
-    spilledBytes = new long[numMetrics]();
-    spilledRows = new long[numMetrics]();
-    spilledFiles = new long[numMetrics]();
-    spilledPartitions = new long[numMetrics]();
-    numDynamicFiltersProduced = new long[numMetrics]();
-    numDynamicFiltersAccepted = new long[numMetrics]();
-    numReplacedWithDynamicFilterRows = new long[numMetrics]();
-    flushRowCount = new long[numMetrics]();
-    scanTime = new long[numMetrics]();
-    skippedSplits = new long[numMetrics]();
-    processedSplits = new long[numMetrics]();
-    skippedStrides = new long[numMetrics]();
-    processedStrides = new long[numMetrics]();
+  unsigned int numMetrics = 0;
+  long veloxToArrow = 0;
+
+  std::unique_ptr<long[]> array; // underlying memory buffer
+  long* arrayRawPtr = nullptr; // point to array.get() after array created
+
+  enum TYPE {
+    kBegin = 0, // begin from 0
+
+    kInputRows = kBegin,
+    kInputVectors,
+    kInputBytes,
+
+    kRawInputRows,
+    kRawInputBytes,
+
+    kOutputRows,
+    kOutputVectors,
+    kOutputBytes,
+
+    // CpuWallTiming.
+    kCpuCount,
+    kWallNanos,
+
+    kPeakMemoryBytes,
+    kNumMemoryAllocations,
+
+    // Spill
+    kSpilledBytes,
+    kSpilledRows,
+    kSpilledPartitions,
+    kSpilledFiles,
+
+    // Runtime metrics.
+    kNumDynamicFiltersProduced,
+    kNumDynamicFiltersAccepted,
+    kNumReplacedWithDynamicFilterRows,
+    kFlushRowCount,
+    kScanTime,
+    kSkippedSplits,
+    kProcessedSplits,
+    kSkippedStrides,
+    kProcessedStrides,
+
+    kEnd,
+    kNum = kEnd - kBegin
```

Comment:
I believe `kNum` should be not an item of this enum because it is not one possible option. Maybe change it to a public variable of `Metrics`.

Reply:
kNum is used to stand for enum item count,  We can use it to determine the length of an array, it's a frequently used style, so let's keep this?

Reply:
OK. Makes sense.